src/browser/connection/index.js

```diff
@@ -0,0 +1,229 @@
+import Promise from 'pinkie';
+import Mustache from 'mustache';
+import { pull as remove } from 'lodash';
+import { parse as parseUserAgent } from 'useragent';
+import { readSync as read } from 'read-file-relative';
+import promisifyEvent from 'promisify-event';
+import shortId from 'shortid';
+import COMMAND from './command';
+import { GeneralError } from '../../errors/runtime';
+import MESSAGE from '../../errors/runtime/message';
+
+
+// Const
+const IDLE_PAGE_TEMPLATE = read('../../client/browser/idle-page/index.html.mustache');
+
+
+var connections = {};
+
+
+export default class BrowserConnection extends EventEmitter {
+    constructor (gateway, provider, browserInfo, permanent) {
+        super();
+
+        this.HEARTBEAT_TIMEOUT = 2 * 60 * 1000;
+
+        this.id                       = BrowserConnection._generateId();
+        this.jobQueue                 = [];
+        this.initScriptsQueue         = [];
+        this.browserConnectionGateway = gateway;
+        this.browserInfo              = browserInfo;
+        this.provider                 = provider;
+
+        this.permanent        = permanent;
+        this.closed           = false;
+        this.ready            = false;
+        this.opened           = false;
+        this.idle             = true;
+        this.switchingToIdle  = false;
+        this.heartbeatTimeout = null;
+
+        this.url           = `${gateway.domain}/browser/connect/${this.id}`;
+        this.heartbeatUrl  = `${gateway.domain}/browser/heartbeat/${this.id}`;
+        this.idleUrl       = `${gateway.domain}/browser/idle/${this.id}`;
+        this.statusUrl     = `${gateway.domain}/browser/status/${this.id}`;
+        this.initScriptUrl = `${gateway.domain}/browser/init-script/${this.id}`;
+
+        this.on('error', () => {
+            this._forceIdle();
+            this.close();
+        });
+
+        connections[this.id] = this;
+
+        this.browserConnectionGateway.startServingConnection(this);
+
+        this._runBrowser();
+    }
+
+    static _generateId () {
+        return shortId.generate();
+    }
+
+    _runBrowser () {
+        // NOTE: Give caller time to assign event listeners
+        process.nextTick(async () => {
+            try {
+                await this.provider.openBrowser(this.id, this.browserInfo, this.url);
+
+                if (!this.ready)
+                    await promisifyEvent(this, 'ready');
+
+                this.opened = true;
+                this.emit('opened');
+            }
+            catch (err) {
+                this.emit('error', new GeneralError(MESSAGE.browserOpeningFailed, this.browserInfo.alias, err.message));
+            }
+        });
+    }
+
+    async _closeBrowser () {
+        if (!this.idle)
+            await promisifyEvent(this, 'idle');
+
+        try {
+            await this.provider.closeBrowser(this.id, { title: this.idleUrl, url: this.idleUrl });
+        }
+        catch (err) {
+            // TODO: Add a warning here?
```

Comment:
Unfortunately, this warning will be emitted after all tasks are done, so it will not be passed to the reporter. Looks like we need a `Runner`-related log in order to implement it.